src/tmpfiles/tmpfiles.c

```diff
@@ -2003,6 +2005,182 @@ static int glob_item_recursively(Item *i, fdaction_t action) {
         return r;
 }
 
+static int rm_if_wrong_type_safe(
+                mode_t mode,
+                int parent_fd,
+                const struct stat *parent_st /* Only used if follow is true. */,
+                const char *name,
+                int flags) {
+        _cleanup_free_ char *parent_name = NULL;
+        bool follow_links = (flags & AT_SYMLINK_NOFOLLOW) == 0;
+        struct stat st;
+        int r;
+
+        assert(name);
+        assert((mode & ~S_IFMT) == 0);
+        assert(!follow_links || parent_st);
+        assert((flags & ~AT_SYMLINK_NOFOLLOW) == 0);
+
+        if (!filename_is_valid(name))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "\"%s\" is not a valid filename.", name);
+
+        if (fstatat(parent_fd, name, &st, flags) < 0) {
+                (void) fd_get_path(parent_fd, &parent_name);
+                return log_full_errno(errno == ENOENT? LOG_DEBUG : LOG_ERR, errno,
+                              "Failed to stat \"%s\" at \"%s\": %m", name, strna(parent_name));
+        }
+
+        /* Fail before removing anything if this is an unsafe transition. */
+        if (follow_links && unsafe_transition(parent_st, &st)) {
+                (void) fd_get_path(parent_fd, &parent_name);
+                return log_error_errno(SYNTHETIC_ERRNO(ENOLINK),
+                                "Unsafe transition from \"%s\" to \"%s\".", parent_name, name);
+        }
+
+        if ((st.st_mode & S_IFMT) == mode)
+                return 0;
+
+        (void) fd_get_path(parent_fd, &parent_name);
+        log_notice("Wrong file type 0x%x; rm -rf \"%s/%s\"", st.st_mode & S_IFMT, strna(parent_name), name);
+
+        /* If the target of the symlink was the wrong type, the link needs to be removed instead of the
+         * target, so make sure it is identified as a link and not a directory. */
+        if (follow_links && fstatat(parent_fd, name, &st, AT_SYMLINK_NOFOLLOW) < 0)
+                return log_error_errno(errno,
+                                "Failed to stat \"%s\" at \"%s\": %m", name, strna(parent_name));
+
+        /* Do not remove mount points. */
+        r = fd_is_mount_point(parent_fd, name, follow_links ? AT_SYMLINK_FOLLOW : 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to check if  \"%s/%s\" is a mountpoint: %m", strna(parent_name), name);
```

Comment:
I'd handle failures of fd_is_mount_point() gracefully, i.e. warn about it, but then assume this is not a mount point. On some old kernels we need some APIs that are privileged and sometime not available, hence it's wise to be lenient here